apollo/client.tsx

```diff
@@ -0,0 +1,153 @@
+            PageComponent.displayName || PageComponent.name || 'Component'
+
+        if (displayName === 'App') {
+            console.warn('This withApollo HOC only works with PageComponents.')
+        }
+
+        WithApollo.displayName = `withApollo(${displayName})`
+    }
+
+    if (ssr || PageComponent.getInitialProps) {
+        WithApollo.getInitialProps = async (ctx) => {
+            const { AppTree } = ctx
+
+            // Initialize ApolloClient, add it to the ctx object so
+            // we can use it in `PageComponent.getInitialProp`.
+            const apolloClient = (ctx.apolloClient = initApolloClient())
+
+            // Run wrapped getInitialProps methods
+            let pageProps = {}
+            if (PageComponent.getInitialProps) {
+                pageProps = await PageComponent.getInitialProps(ctx)
+            }
+
+            // Only on the server:
+            if (typeof window === 'undefined') {
+                // When redirecting, the response is finished.
+                // No point in continuing to render
+                if (ctx.res && ctx.res.finished) {
+                    return pageProps
+                }
+
+                // Only if ssr is enabled
+                if (ssr) {
+                    try {
+                        // Run all GraphQL queries
+                        const { getDataFromTree } = await import(
+                            '@apollo/react-ssr'
+                        )
+                        await getDataFromTree(
+                            <AppTree
+                                pageProps={{
+                                    ...pageProps,
+                                    apolloClient,
+                                }}
+                            />
+                        )
+                    } catch (error) {
+                        // Prevent Apollo Client GraphQL errors from crashing SSR.
+                        // Handle them in components via the data.error prop:
+                        // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
+                        console.error(
+                            'Error while running `getDataFromTree`',
+                            error
+                        )
+                    }
+
+                    // getDataFromTree does not call componentWillUnmount
+                    // head side effect therefore need to be cleared manually
+                    Head.rewind()
+                }
+            }
+
+            // Extract query data from the Apollo store
+            const apolloState = apolloClient!.cache.extract()
+
+            return {
+                ...pageProps,
+                apolloState,
+            }
+        }
+    }
+
+    return WithApollo
+}
+
+/**
+ * Always creates a new apollo client on the server
+ * Creates or reuses apollo client in the browser.
+ * @param  {Object} initialState
+ */
+export function initApolloClient(initialState) {
```

Comment:
How about documenting and/or adding a default parameter to the optional `initialState` arg, because there are a few places where this functions is invouked without passing an argument, which breakes the `TS2554` rule (`Expected X arguments, but got 0.`)

Reply:
Actually, I think we should let developers mkdir manually. Now I changed my mind, added a script to do this.

The type issue was also fixed.